active/0004-extensibility-support-for-deployment-stacks.md

```diff
@@ -0,0 +1,405 @@
+
+```bicep
+@secure()
+param kubeConfigParam string
+
+extension kubernetes with { // this will produce a diagnostic if the config is not provided
+  kubeConfig: kubeConfigParam
+  namespace: 'default'
+}
+
+// ... extension resources
+```
+
+In this example, the extension configuration contains a secret from a deployment parameter. A deployment parameter can
+be supplied by the user or sourced from a parent deployment. In either case, it is a user secret and thus cannot be 
+persisted long term by stacks for use in future resource deletion. 
+
+In the case where it's supplied by the user and without secret persistence, the only way the stacks service could 
+possibly source this data again is to prompt the user for it. It is unclear how that experience would work. Does the 
+user need to supply the entire old deployment's parameters back to the stack service? How do they know which version of 
+inputs to supply? What if a subset of those parameters are no longer relevant? Would they only need to supply parameters
+used in configurations? How would we detect and differentiate that across an entire tree of deployments?
+
+In the case where it's supplied by a parent deployment, the problem becomes recursive. It's either provided by the user,
+is sourced from somewhere else in that parent deployment (such as another resource's outputs like a key vault or AKS 
+cluster), or sourced from another parent deployment.
+
+To complicate this further, any pathway between the secret's source and consumption can have dynamic language expressions
+such as template function calls or branching. This would effectively require stacks to generate a separate deployment template
+at deletion time from the user's deployment template to process these language expressions.
+
+The stacks service cannot store user secrets, and it should not evaluate dynamic deployment template language expressions.
+The stacks service needs extension configurations to be defined in a format that is interpretable at deletion time that
+is transparent to the user and secure.
+
+### The new design from the Bicep perspective
+
+The following Bicep code illustrates the new design for extension configurations. The root deployment supplies the
+extension configuration details to the child deployment, allowing direct access to ARM resources in the root deployment
+that supply secrets.
+
+main.bicep - The root deployment
+```bicep
+// The ARM aksCluster resource can supply kubeConfigs for the AKS Kubernetes use case.
+resource aksCluster '...' = {
+  // ...
+}
+
+module extResources 'extResources.bicep' = {
+  name: 'extResourcesDeploymentName'
+  extensionConfigs: {
+    k8s: {
+      kubeConfig: aksCluster.listClusterAdminCredential().kubeconfig[0].value
+      namespace: 'default'
+    }
+  }
+  params: {
+    // ...
+  }
+}
+```
+
+extResources.bicep - Child deployment with extension resources
+```bicep
+extension kubernetes as k8s // optional alias
```

Comment:
Could you cover what the UX looks like for child -> grandchild module? 

E.g. it would be nice to avoid having to write the following if we're able to detect and automatically flow down the boilerplate:

```bicep
extension kubernetes

module foo 'foo.bicep' = {
  extensionConfigs: {
    kubernetes: kubernetes
  }
}
```

/cc @rynowak

Reply:
Good point. This will need to be implemented because passing the data as deployment parameters   would be non-sensical.

Reply:
Yes please. Extension config will be much easier to use if it automatically flows across module boundaries. 

We implemented this in the Radius DE fork and have been living with it happily for a while.

Reply:
I expanded on this in recent commits. There will need to be an "inherit" mechanism for "extensionConfigs" in the ARM template level. I believe the substitution should be able to be done within the DeploymentResourceJob for deployment resource types as that has access to the deployment entity + child deployment properties.